src/pybind/mgr/dashboard_v2/tools.py

```diff
@@ -347,3 +349,89 @@ def _callback(self):
             if name in cookie:
                 del cookie[name]['expires']
                 del cookie[name]['max-age']
+
+
+class NotificationQueue(threading.Thread):
+    _ALL_TYPES_ = '__ALL__'
+    _listeners = collections.defaultdict(lambda: [])
+    _lock = threading.Lock()
+    _cond = threading.Condition()
+    _queue = collections.deque()
+    _running = False
+
+    def __init__(self):
+        super(NotificationQueue, self).__init__()
+
+    @classmethod
+    def start_queue(cls):
+        cls._running = True
+        cls._instance = NotificationQueue()
+        cls._instance.start()
+
+    @classmethod
+    def stop(cls):
+        cls._running = False
+        with cls._cond:
+            cls._cond.notify()
+        cls._instance.join()
+
+    @classmethod
+    def register(cls, func, types=None):
+        """Registers function to listen for notifications
+
+        If the second parameter `types` is ommited, the function in `func`
+        parameter will be called for any type of notifications.
+
+        Args:
+            func (function): python function ex: def foo(val)
+            types (str|list): the single type to listen, or a list of types
+        """
+        with cls._lock:
+            if not types:
+                if func in cls._listeners[cls._ALL_TYPES_]:
+                    return
+                cls._listeners[cls._ALL_TYPES_].append(func)
+                return
+            if isinstance(types, str):
+                if func in cls._listeners[types]:
+                    return
+                cls._listeners[types].append(func)
+            elif isinstance(types, list):
+                for typ in types:
+                    if func in cls._listeners[typ]:
+                        continue
+                    cls._listeners[typ].append(func)
+            else:
+                raise Exception("types param is neither a string nor a list")
+
+    @classmethod
+    def new_notification(cls, notify_type, notify_value):
+        cls._queue.append((notify_type, notify_value))
+        with cls._cond:
+            cls._cond.notify()
+
+    @classmethod
+    def notify_listeners(cls, events):
+        if events:
```

Comment:
`if events` is superfluous. Otherwise, please make it obvious that `events` can be `None`.